### PR TITLE
Allow restarting groups migration after backup groups were created and workspace-local groups deleted

### PR DESCRIPTION
**Pull request by mwojtyczka**
Fix for issue [386](https://github.com/databrickslabs/ucx/issues/386)

### CODE REVIEW

File: src/databricks/labs/ucx/workspace_access/groups.py [hunk 1]
```diff
@@ -117,18 +117,22 @@ def _get_or_create_backup_group(self, source_group_name: str, source_group: iam.
...
             ws_group = self._get_group(name, "workspace")
-            assert ws_group, f"Group {name} not found on the workspace level"
+            if ws_group is None:  # this can happen upon restart of the tool when backup group has already been created
+                logger.info(f"Group {name} not found on the workspace level but backup group exists")
+
             acc_group = self._get_group(name, "account")
             assert acc_group, f"Group {name} not found on the account level"
+
             backup_group = self._get_or_create_backup_group(source_group_name=name, source_group=ws_group)
+
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/groups.py:135]: remove blank line here

File: src/databricks/labs/ucx/workspace_access/groups.py [hunk 2]
```diff
@@ -117,18 +117,22 @@ def _get_or_create_backup_group(self, source_group_name: str, source_group: iam.
...
     def _set_migration_groups(self, groups_names: list[str]):
         def get_group_info(name: str):
             ws_group = self._get_group(name, "workspace")
-            assert ws_group, f"Group {name} not found on the workspace level"
+            if ws_group is None:  # this can happen upon restart of the tool when backup group has already been created
+                logger.info(f"Group {name} not found on the workspace level but backup group exists")
+
             acc_group = self._get_group(name, "account")
             assert acc_group, f"Group {name} not found on the account level"
+
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/groups.py:133]: remove blank line here

File: src/databricks/labs/ucx/workspace_access/groups.py [hunk 3]
```diff
@@ -117,18 +117,22 @@ def _get_or_create_backup_group(self, source_group_name: str, source_group: iam.
             roles=source_group.roles,
             members=source_group.members,
         )
-        self._workspace_groups.append(backup_group)
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/groups.py:120]: revert this removal

File: src/databricks/labs/ucx/workspace_access/groups.py [hunk 4]
```diff
@@ -237,15 +250,10 @@ def has_groups(self) -> bool:
...
     def migration_groups_provider(self) -> GroupMigrationState:
-        if len(self._migration_state.groups) == 0:
-            logger.info("No groups were loaded or initialized, nothing to do")
         return self._migration_state
 
     def replace_workspace_groups_with_account_groups(self):
         logger.info("Replacing the workspace groups with account-level groups")
-        if len(self._migration_state.groups) == 0:
-            logger.info("No groups were loaded or initialized, nothing to do")
-            return True
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/groups.py:248]: revert this removal

File: src/databricks/labs/ucx/workspace_access/groups.py [hunk 5]
```diff
@@ -237,15 +250,10 @@ def has_groups(self) -> bool:
 
     @property
     def migration_groups_provider(self) -> GroupMigrationState:
-        if len(self._migration_state.groups) == 0:
-            logger.info("No groups were loaded or initialized, nothing to do")
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/groups.py:241]: revert this removal

File: tests/unit/workspace_access/test_groups.py [hunk 6]
```diff
@@ -186,8 +186,7 @@ def test_prepare_groups_in_environment_with_one_group_in_conf_should_return_migr
     manager.prepare_groups_in_environment()
 
     group_info = MigrationGroupInfo(workspace=ws_de_group, account=acc_de_group, backup=backup_de_group)
-    assert manager._migration_state.groups == [group_info]
-    assert len(manager._workspace_groups) == 2  # created backup group should be added to the list
+    assert manager.migration_groups_provider.groups == [group_info]
```
Comment (nfx) [inline tests/unit/workspace_access/test_groups.py:189]: don't change the existing tests

File: tests/unit/workspace_access/test_groups.py [hunk 7]
```diff
@@ -213,7 +212,7 @@ def test_prepare_groups_in_environment_with_multiple_groups_in_conf_should_retur
     de_group_info = MigrationGroupInfo(workspace=ws_de_group, account=acc_de_group, backup=backup_de_group)
     ds_group_info = MigrationGroupInfo(workspace=ws_ds_group, account=acc_ds_group, backup=backup_ds_group)
 
-    assert compare(manager._migration_state.groups, [ds_group_info, de_group_info])
+    assert compare(manager.migration_groups_provider.groups, [ds_group_info, de_group_info])
```
Comment (nfx) [inline tests/unit/workspace_access/test_groups.py:215]: don't change the existing tests

File: tests/unit/workspace_access/test_groups.py [hunk 8]
```diff
@@ -260,7 +259,7 @@ def test_prepare_groups_in_environment_with_backup_group_not_created_should_crea
     manager = GroupManager(client, group_conf)
     manager.prepare_groups_in_environment()
     group_info = MigrationGroupInfo(workspace=ws_de_group, account=acc_de_group, backup=backup_de_group)
-    assert manager._migration_state.groups == [group_info]
+    assert manager.migration_groups_provider.groups == [group_info]
```
Comment (nfx) [inline tests/unit/workspace_access/test_groups.py:262]: don't change the existing tests

File: tests/unit/workspace_access/test_groups.py [hunk 9]
```diff
@@ -282,7 +281,7 @@ def test_prepare_groups_in_environment_with_conf_in_auto_mode_should_populate_mi
     manager.prepare_groups_in_environment()
 
     group_info = MigrationGroupInfo(workspace=ws_de_group, account=acc_de_group, backup=backup_de_group)
-    assert manager._migration_state.groups == [group_info]
+    assert manager.migration_groups_provider.groups == [group_info]
```
Comment (nfx) [inline tests/unit/workspace_access/test_groups.py:284]: don't change the existing tests

File: tests/unit/workspace_access/test_groups.py [hunk 10]
```diff
@@ -347,7 +360,7 @@ def test_system_groups():
     group_conf = GroupsConfig(backup_group_prefix="dbr_backup_", selected=["admins"])
     manager = GroupManager(client, group_conf)
     manager.prepare_groups_in_environment()
-    assert len(manager._migration_state.groups) == 0
+    assert len(manager.migration_groups_provider.groups) == 0
```
Comment (nfx) [inline tests/unit/workspace_access/test_groups.py:363]: don't change the existing tests

File: tests/unit/workspace_access/test_groups.py [hunk 11]
```diff
@@ -366,7 +379,7 @@ def test_workspace_only_groups():
     group_conf = GroupsConfig(backup_group_prefix="dbr_backup_", selected=["ws_group"])
     manager = GroupManager(client, group_conf)
     manager.prepare_groups_in_environment()
-    assert len(manager._migration_state.groups) == 0
+    assert len(manager.migration_groups_provider.groups) == 0
```
Comment (nfx) [inline tests/unit/workspace_access/test_groups.py:382]: don't change the existing tests

**Review by nfx — CHANGES_REQUESTED** [review]
see comments. also i'd require an integration test for this, otherwise we're not sure this fix actually works.

**Comment by mwojtyczka** [comment]
> see comments. also i'd require an integration test for this, otherwise we're not sure this fix actually works.

I agree, I will add integration tests.

**Comment by mwojtyczka** [comment]
won't do, as the functionality is already covered with different PR

**Pull request closed without merging**